pkg/cluster/cluster.go

```diff
@@ -91,20 +91,37 @@ func NewPeer(
 	}
 	p.mlist = ml
 
-	n, _ := ml.Join(knownPeers)
-	level.Debug(l).Log("msg", "joined cluster", "peers", n)
+	n, err := ml.Join(joinCfg.KnownPeers)
+	if err != nil {
+		return nil, errors.Wrap(err, "join cluster")
+	}
 
+	level.Debug(l).Log("msg", "joined cluster", "peers", n)
 	if n > 0 {
 		go p.warnIfAlone(l, 10*time.Second)
 	}
+
 	// Initialize state with ourselves.
 	p.mtx.RLock()
 	defer p.mtx.RUnlock()
-	p.data[p.Name()] = PeerState{Type: typ, APIAddr: apiAddr}
+
+	p.data[p.Name()] = state
 
 	return p, nil
 }
 
+// SetLabels sets the labels and notifies other members.
+func (p *Peer) SetLabels(labels []storepb.Label, timeout time.Duration) error {
```

Comment:
Where is this being used?

Reply:
.. it should be used after querying correct labels. Will add that (:

Reply:
I'm a bit concerned about the timeout. The purpose of the gossip layer is ultimately that we can distribute state in an eventually consistent way without worrying about it reaching peers etc.

Adding a timeout and error here re-surfaces this concern. What's the appropriate timeout? It failed, now what? Does the sidecar retry indefinitely, or for 10 minutes?

In the end the gossip layer does all this for us. I'd just set the internal label state and call it a day. I'm also not quite sure that `UpdateNode` actually does anything for our custom state. The meta info it refers to is probably the 512 byte slice of Node which we don't use: https://godoc.org/github.com/hashicorp/memberlist#TransmitLimitedQueue.QueueBroadcast

Reply:
hm.. yeah. And we need to have some test for that as well. 

Ok let me remove SetLabel from this PR. This PR is starting touching too many things.